src/raftstore/store/peer.rs

```diff
@@ -1882,6 +1807,111 @@ impl Peer {
     }
 }
 
+/// `RequestPolicy` decides how we handle a request.
+#[derive(Clone, PartialEq, Debug)]
+enum RequestPolicy {
+    // Handle the read request directly without dispatch.
+    ReadLocal,
+    // Handle the read request via raft's SafeReadIndex mechanism.
+    ReadIndex,
+    ProposeNormal,
+    ProposeTransferLeader,
+    ProposeConfChange,
+}
+
+/// `RequestInspector` makes `RequestPolicy` for requests.
+trait RequestInspector {
+    /// Has the current term been applied?
+    fn has_applied_current_term(&self) -> bool;
+    /// Inspects its lease.
+    fn inspect_lease(&mut self) -> LeaseState;
+
+    /// Inspect a request, return a policy that tells us how to
+    /// handle the request.
+    fn inspect(&mut self, req: &RaftCmdRequest) -> Result<RequestPolicy> {
+        if req.has_admin_request() {
+            if apply::get_change_peer_cmd(req).is_some() {
+                return Ok(RequestPolicy::ProposeConfChange);
+            }
+            if get_transfer_leader_cmd(req).is_some() {
+                return Ok(RequestPolicy::ProposeTransferLeader);
+            }
+            return Ok(RequestPolicy::ProposeNormal);
+        }
+
+        let mut has_read = false;
+        let mut has_write = false;
+        for r in req.get_requests() {
+            match r.get_cmd_type() {
+                CmdType::Get | CmdType::Snap => has_read = true,
+                CmdType::Delete | CmdType::Put | CmdType::DeleteRange | CmdType::IngestSST => {
+                    has_write = true
+                }
+                CmdType::Prewrite | CmdType::Invalid => {
+                    return Err(box_err!(
+                        "invalid cmd type {:?}, message maybe currupted",
+                        r.get_cmd_type()
+                    ));
+                }
+            }
+
+            if has_read && has_write {
+                return Err(box_err!("read and write can't be mixed in one batch."));
+            }
+        }
+
+        if has_write {
+            return Ok(RequestPolicy::ProposeNormal);
+        }
+
+        if req.get_header().get_read_quorum() {
+            return Ok(RequestPolicy::ReadIndex);
+        }
+
+        // If applied index's term is differ from current raft's term, leader transfer
+        // must happened, if read locally, we may read old value.
+        if !self.has_applied_current_term() {
+            return Ok(RequestPolicy::ReadIndex);
+        }
+
+        // Local read should be performed, if and only if leader is in lease.
+        // None for now.
+        match self.inspect_lease() {
+            LeaseState::Valid => Ok(RequestPolicy::ReadLocal),
+            LeaseState::Expired | LeaseState::Suspect => {
+                // Perform a consistent read to Raft quorum and try to renew the leader lease.
+                Ok(RequestPolicy::ReadIndex)
+            }
+        }
+    }
+}
+
+impl RequestInspector for Peer {
+    fn has_applied_current_term(&self) -> bool {
```

Comment:
It may not be accurate. As it has not applied all logs from current term.

Reply:
What's your suggestion?

Reply:
How about `has_applied_to_current_term`?